diff --git a/Cargo.toml b/Cargo.toml
index dbecd26..89bc1e8 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -15,7 +15,7 @@ repository = "https://github.com/Chik-Network/chik_rs/"
 serde = { version = "1.0.130", features = ["derive"] }
 klvmr = "=0.1.20"
 pyo3 = "0.15.1"
-bincode = { git = "https://github.com/richardkiss/bincode", branch = "chik" }
+# bincode = { git = "https://github.com/richardkiss/bincode", branch = "chik" }
 hex = "=0.4.3"
 
 [dev-dependencies]
diff --git a/src/lib.rs b/src/lib.rs
index 759f6d4..f4fa654 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1,2 +1,2 @@
 pub mod gen;
-pub mod streamable;
+// pub mod streamable;

diff --git a/wheel/src/api.rs b/wheel/src/api.rs
index 6e34218..a096e84 100644
--- a/wheel/src/api.rs
+++ b/wheel/src/api.rs
@@ -2,8 +2,8 @@ use crate::run_generator::__pyo3_get_function_run_generator;
 use chik::gen::flags::COND_ARGS_NIL;
 use chik::gen::flags::COND_CANON_INTS;
 use chik::gen::flags::NO_UNKNOWN_CONDS;
-use chik::streamable::coin::Coin;
-use chik::streamable::fullblock::Fullblock;
+//use chik::streamable::coin::Coin;
+//use chik::streamable::fullblock::Fullblock;
 use klvmr::chik_dialect::NO_NEG_DIV;
 use klvmr::chik_dialect::NO_UNKNOWN_OPS;
 use klvmr::py::lazy_node::LazyNode;
@@ -25,8 +25,8 @@ pub fn chik_rs(_py: Python, m: &PyModule) -> PyResult<()> {
     m.add("NO_UNKNOWN_OPS", NO_UNKNOWN_OPS)?;
     m.add("MEMPOOL_MODE", MEMPOOL_MODE)?;
     m.add_class::<LazyNode>()?;
-    m.add_class::<Coin>()?;
-    m.add_class::<Fullblock>()?;
+    //m.add_class::<Coin>()?;
+    //m.add_class::<Fullblock>()?;
 
     Ok(())
 }
